src/client/handlers/VaultsSecretsMkdir.ts

```diff
@@ -2,50 +2,60 @@ import type { DB } from '@matrixai/db';
 import type {
   ClientRPCRequestParams,
   ClientRPCResponseResult,
-  SecretMkdirMessage,
-  SuccessMessage,
+  SecretDirMessage,
+  SuccessWithErrorMessage,
 } from '../types';
 import type VaultManager from '../../vaults/VaultManager';
-import { UnaryHandler } from '@matrixai/rpc';
+import { DuplexHandler } from '@matrixai/rpc';
 import * as vaultsUtils from '../../vaults/utils';
 import * as vaultsErrors from '../../vaults/errors';
 import * as vaultOps from '../../vaults/VaultOps';
 
-class VaultsSecretsMkdir extends UnaryHandler<
+class VaultsSecretsMkdir extends DuplexHandler<
   {
     vaultManager: VaultManager;
     db: DB;
   },
-  ClientRPCRequestParams<SecretMkdirMessage>,
-  ClientRPCResponseResult<SuccessMessage>
+  ClientRPCRequestParams<SecretDirMessage>,
+  ClientRPCResponseResult<SuccessWithErrorMessage>
 > {
-  public handle = async (
-    input: ClientRPCRequestParams<SecretMkdirMessage>,
-  ): Promise<ClientRPCResponseResult<SuccessMessage>> => {
-    const { vaultManager, db } = this.container;
-    await db.withTransactionF(async (tran) => {
-      const vaultIdFromName = await vaultManager.getVaultId(
-        input.nameOrId,
-        tran,
-      );
-      const vaultId =
-        vaultIdFromName ?? vaultsUtils.decodeVaultId(input.nameOrId);
-      if (vaultId == null) {
-        throw new vaultsErrors.ErrorVaultsVaultUndefined();
-      }
-      await vaultManager.withVaults(
-        [vaultId],
-        async (vault) => {
-          await vaultOps.mkdir(vault, input.dirName, {
-            recursive: input.recursive,
-          });
-        },
-        tran,
-      );
-    });
-    return {
-      success: true,
-    };
+  public handle = async function* (
+    input: AsyncIterable<ClientRPCRequestParams<SecretDirMessage>>,
+  ): AsyncGenerator<ClientRPCResponseResult<SuccessWithErrorMessage>> {
+    const { vaultManager, db }: { vaultManager: VaultManager; db: DB } =
+      this.container;
```

Comment:
I had to do this as my LSP couldn't detect the types of the values inside `this.container`. The issue can be observed in the `VaultsSecretsGet` handler, which is a `DuplexStream` and suffers with the same issue of the types not being recognised. This interestingly doesn't happen with `UnaryStream` or `ServerStream`, only in `DuplexStream` and `ClientStream`.

This is important as without the types being recognised, objects like `vaultManager` has the `any` type and no autocomplete support exists. It is also really hard to write correct code, as we could be relying on a missing property/attribute, but because the default type is `any`, no errors will be shown while writing the code.

If this issue is not reproducible (eg. JetBrains or VSCode can infer the type properly), then should adding types like this be a standard for all commands, or should this be excluded as it is not a big enough issue?

If this issue can be reproduced (same behaviour on other IDEs), then should this be looked into, or do we use this approach to properly define types for injected dependencies, or do we ignore the issue completely?

Reply:
From what I recall, anonymous generators get weird with `this`. Look at other handlers like `VaultsSecretsList` that seems to be how to handle it. Make sure all the handlers that function like generators are consistent in this.

Reply:
`VaultsSecretsList` is a `ServerHandler` and doesn't have this issue, even though it is an anonymous generator. So, it only applies when we are getting an input stream. For consistency, I can do this to all the `VaultsSecrets` command, but I think at least we should try to understand why this could be happening.

Reply:
That doesn't make sense.

I just had a look at it and it is a normal generator function, not a anonymous one. That alone makes it inconsistent with the other handlers.

Reply:
Yeah, I see that now. The functions which are written in the first way have full type hinting, and the ones written with the second approach don't.

I will make sure all `VaultsSecrets` handlers use the `public handle = async function* ()` syntax. I will also use the type definition for unpacking the container to get type hints, too.

I will resolve this when I finish implementing this.

Reply:
Should I go over all the places where we use the first approach and switch it out for the second one, or only in the client domain's handlers? Or only in `VaultsSecrets` handlers? What is the scope of this change?

Reply:
I have discovered some issues with the code while doing this change.

In `src/client/handlers/NodesGetAll.ts`, we pass in `nodeGraph` and `keyRing` into the container, but only take out `nodeGraph`. Why do we have `keyRing` if we are not going to use it?

Reply:
If we don't use it there then remove it.